### ISSUE TITLE
[Community Pipelines]

### ISSUE THREAD

**Issue by NatalieC323** [hpcaitech/ColossalAI#3487]
## Community Pipelines
---
### Description:

We are thrilled to announce the latest updates to ColossalChat, an open-source solution for cloning ChatGPT with a complete RLHF (Reinforcement Learning with Human Feedback) pipeline. 

As Colossal-AI undergoes major updates, we are actively maintaining ColossalChat to stay aligned with the project's progress. With the introduction of Community-driven Pipelines, we aim to create a collaborative platform for developers to contribute exotic features built on top of ColossalChat.

Community-driven Pipelines is an initiative that allows users to contribute their own pipelines to the ColossalChat package, fostering a sense of community and making it easy for others to access and benefit from shared work. The primary goal with community-driven pipelines is to have a community-maintained collection of diverse and exotic functionalities built on top of the ColossalChat package, which is powered by the Colossal-AI project and its Coati module (ColossalAI Talking Intelligence).

### How to use community-driven features:

Load community-driven pipelines by passing the appropriate argument(s) to the Coati class or any other relevant class, as specified in the examples/community folder. Here's an example of how to use a community-driven pipeline:

```
model_name = "Coati7B"  # Replace with the actual model name
custom_pipeline_name = "CustomPipeline"  # Replace with the filename of the custom pipeline in the 'community_pipelines' folder
chat_instance = ColossalChat.from_pretrained(model_name, custom_pipeline=custom_pipeline_name)
```

Users can load custom pipelines from the `community_pipelines` folder using the code similar to the one you provided. Contribute to the community-driven features by sending a PR with your own implementations to the examples/community folder, and we will merge them promptly.

### Why community-driven features?
While the code for community-driven features may not be included in the official PyPI releases, it is usable with the ColossalChat package >= X.X.X. The reason for this is:

- There is an extensive range of use cases that could be supported, but it's challenging to determine which are important enough to be rigorously maintained.
- Different opinions on the design philosophy of various features.
- We want to encourage the community to add features they would like to contribute, but at the same time, we cannot take the burden of full maintenance upon ourselves.

By providing community-driven pipelines, we allow community members to contribute and share their work while enjoying flexibility and development speed, making their work easily accessible to the rest of the ecosystem.

### What pipelines to contribute?

Feel free to contribute any pipeline that you find exciting! If you need some ideas. We are particularly interested in community contributions in the following areas:

1. Developing new features that leverage ColossalChat's capabilities.

2. Enhancing existing functionalities to improve performance and user experience.

3. Identifying and addressing limitations of the Coati7B model.

4. Expanding the range of use cases and examples for Coati7B.

If you want some ideas, we've created some ideas for different issues here, but you can do other pipelines as well! Some examples (#XXX for possible issue number):

- [ ] [Community Pipelines] Custom dataset integration #XXX
- [ ] [Community Pipelines] Alternative reward models #XXX
- [ ] [Community Pipelines] Customized RLHF algorithms #XXX
- [ ] [Community Pipelines] Integration with external RL libraries #XXX
- [ ] [Community Pipelines] Additional quantization techniques #XXX
- [ ] [Community Pipelines] Model deployment in specific environments #XXX

By providing community-driven features, we allow community members to contribute and share their work while enjoying the flexibility and development speed, making their work easily accessible to the rest of the ecosystem.

### How to contribute to a community pipeline?

To contribute to a community pipeline, please create a file or folder under the examples/community directory that contains your implementation. Make sure your code is well-documented and follows the recommended coding guidelines. Once ready, submit a PR to the ColossalChat repository, and we will review and merge it accordingly.

Let's create a similar custom pipeline for ColossalChat. We'll create a pipeline that takes a ColossalChat model and an additional component as arguments. Then, we'll define the `__call__ `method to utilize the ColossalChat model and the additional component in a single forward pass.

First, let's define the custom pipeline class inheriting from `ColossalChat`:

from ColossalChat import ColossalChat
import torch

```
class CustomPipeline(ColossalChat):

    def __init__(self, model_name, additional_component):
        super().__init__(model_name)
        
        self.additional_component = additional_component
 
        self.register_modules
(additional_component=additional_component)
```
Now, let's define the __call__ method for our pipeline:

```
import torch

class CustomPipeline(ColossalChat):

    def __init__(self, model_name, additional_component):
        super().__init__(model_name)
        
        self.additional_component = additional_component
        self.register_modules(additional_component=additional_component)

    def __call__(self, input_text):
        chat_output = self.generate_response(input_text)
        additional_output = self.additional_component.process(chat_output)
        
        return additional_output

```
To use this custom pipeline, users can create an instance of the CustomPipeline class, passing the model name and the additional component:

```
from additional_component import AdditionalComponent

model_name = "example/colossal-chat-model"
additional_component = AdditionalComponent()

pipeline = CustomPipeline(model_name=model_name, additional_component=additional_component)

input_text = "Hello, ColossalChat!"
output = pipeline(input_text)
```

To load pre-existing weights and share the custom pipeline with the community, you can create a similar pipeline-loading function:

```
from additional_component import AdditionalComponent

class CustomPipeline(ColossalChat):

    # ... (previous code)

    @classmethod
    def from_pretrained(cls, model_name, custom_pipeline_path):
        additional_component = AdditionalComponent.from_pretrained(custom_pipeline_path)
        return cls(model_name=model_name, additional_component=additional_component)
```

Now users can load the pipeline with pre-existing weights:
### How to Get Involved:

```
pipeline = CustomPipeline.from_pretrained("example/colossal-chat-model", custom_pipeline_path="path/to/additional_component")
output = pipeline(input_text)
```
To share this custom pipeline with the community, you can submit a pull request to add the CustomPipeline code to the `community` folder in the [ColossalChat repository](https://github.com/hpcaitech/ColossalAI/tree/main/applications/Chat/examples/community). Once the pull request is merged, other users can load the custom pipeline using the code provided in the previous answers.

The general steps can be summarized as:
1. Develop your custom pipeline by inheriting from the `ColossalChat` class and implementing the `__init__` and `__call__` methods as shown in the provided examples.

2. Test your custom pipeline to ensure it works correctly with the ColossalChat framework.

3. Create a new Python file for your custom pipeline and place it in the `community_pipelines` folder.

4. Submit a pull request to add your custom pipeline code to the `community_pipelines` folder in the ColossalChat repository. Please include a brief description of your custom pipeline and its features.

Once your pull request is merged, other users can load your custom pipeline using the `from_pretrained` method in the ColossalChat class. Thank you for contributing to ColossalChat!

### How to get involved
To join our community-driven initiative, please visit the [ColossalChat GitHub repository](https://github.com/hpcaitech/ColossalAI/tree/main/applications/Chat/examples), review the provided information, and explore the codebase. To contribute, create a new issue outlining your proposed feature or enhancement, and our team will review and provide feedback. We look forward to collaborating with you on this exciting project!
